### PR TITLE
[release-4.13]  SDN-4186,OCPBUGS-34669: introduce validating webhook

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
This PR manually backports the validating webhook for ovnkube-node to ovn-kubernetes 4.13

including code from PRs mentioned in https://issues.redhat.com/browse/SDN-4186
https://github.com/openshift/cluster-network-operator/pull/1983
https://github.com/openshift/cluster-network-operator/pull/2051

this will fail CI until https://github.com/openshift/ovn-kubernetes/pull/2053 merges

### CODE REVIEW

**Comment by asood-rh** [comment]
Pre merge testing results in JIRA

/ocpbugs cc-qa
/label qe-approved
/label cherry-pick-approved

**Comment by JacobTanenbaum** [comment]
the e2e-aws-ovn-windows test requires the windows-machine-config-operator to update submodules, that should happen by EOD today (6/3/2024)

**Comment by JacobTanenbaum** [comment]
/retest-required

the previous test did not have the newly merged version of wmco

**Comment by JacobTanenbaum** [comment]
/retest
looks like there was some setup issues

File: pkg/network/node_identity.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		data.Data["K8S_LOCAL_APISERVER"] = "https://" + net.JoinHostPort(localAPIServer.Host, localAPIServer.Port)
+
+		webhookDeployment := &appsv1.Deployment{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Deployment",
+				APIVersion: appsv1.SchemeGroupVersion.String(),
+			},
+		}
+		nsn := types.NamespacedName{Namespace: hcpCfg.Namespace, Name: "network-node-identity"}
+		if err := client.Default().CRClient().Get(context.TODO(), nsn, webhookDeployment); err != nil {
```
Comment (kyrtapz) [inline pkg/network/node_identity.go:102]: We are missing https://github.com/openshift/cluster-network-operator/commit/8a09adfea9e5c08ac2c7e4ea2b0182b802ac9db5

File: pkg/network/node_identity.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+			if !apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("failed to retrieve existing network-node-identity deployment: %w", err)
+			} else {
+				klog.Infof("network-node-identity deployment does not exist")
+			}
+		}
```
Comment (kyrtapz) [inline pkg/network/node_identity.go:108]: What happened to https://github.com/openshift/cluster-network-operator/blob/release-4.14/pkg/network/node_identity.go#L114-L116?

File: bindata/network/ovn-kubernetes/common/002-rbac.yaml [hunk 3]
```diff
@@ -24,16 +24,27 @@ rules:
 apiVersion: rbac.authorization.k8s.io/v1
```
Comment (kyrtapz) [inline bindata/network/ovn-kubernetes/common/002-rbac.yaml:24]: ovnkube-node no longer needs access to patch `nodes` and `pods` right? the RBAC still has it:
https://github.com/openshift/cluster-network-operator/blob/f68d528042f49bfd91f3a07588c5cea656cdbffd/bindata/network/ovn-kubernetes/common/002-rbac.yaml#L100-L108
https://github.com/openshift/cluster-network-operator/blob/f68d528042f49bfd91f3a07588c5cea656cdbffd/bindata/network/ovn-kubernetes/common/002-rbac.yaml#L55-L63

**Review by kyrtapz — CHANGES_REQUESTED** [review]

File: pkg/network/node_identity.go (hunk 2 above)
Comment (JacobTanenbaum) [reply pkg/network/node_identity.go:113]: the function `deploymentProgressing()` does not exist

File: pkg/network/node_identity.go (hunk 1 above)
Comment (JacobTanenbaum) [reply pkg/network/node_identity.go]: added

File: pkg/platform/platform.go [hunk 4]
```diff
@@ -14,13 +14,34 @@ import (
...
 
 var cloudProviderConfig = types.NamespacedName{
 	Namespace: "openshift-config-managed",
 	Name:      "kube-cloud-config",
 }
 
+// isNetworkNodeIdentityEnabled determines if network node identity should be enabled.
+// It checks the `enabled` key in the network-node-identity/openshift-network-operator configmap.
+// If the configmap doesn't exist, it returns true (the feature is enabled by default).
+func isNetworkNodeIdentityEnabled(client cnoclient.Client) (bool, error) {
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply pkg/platform/platform.go]: added

File: bindata/network/node-identity/self-hosted/node-identity-webhook.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      caBundle: {{.NetworkNodeIdentityCABundle}}
+    admissionReviewVersions: ['v1']
+    sideEffects: None
+    rules:
+      - operations: [ "UPDATE" ]
+        apiGroups: ["*"]
+        apiVersions: ["*"]
+        resources: ["nodes/status"]
+        scope: "*"
+### ONCE THE pod-admission work is backported to multus we can work on adding the pod-admission
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply bindata/network/node-identity/self-hosted/node-identity-webhook.yaml]: removed the commit as I don't remember and we can see what happens in CI

File: bindata/network/node-identity/self-hosted/node-identity.yaml [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+          secret:
+            secretName: network-node-identity-cert
+        - name: env-overrides
+          configMap:
+            name: env-overrides
+            optional: true
+        - hostPath:
+            path: /var/log/kube-apiserver
+          name: audit-dir
+      tolerations:
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply bindata/network/node-identity/self-hosted/node-identity.yaml:121]: added c7b959f33ee9f3e97bf97b119cc3ab9856b1702f

File: bindata/network/node-identity/managed/node-identity.yaml [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                --webhook-cert-dir=/etc/webhook-cert \
+                --webhook-host="" \
+                --webhook-port={{.NetworkNodeIdentityPort}} \
+                ${ho_enable} \
+                --disable-approver \
+                --extra-allowed-user="system:serviceaccount:openshift-ovn-kubernetes:ovn-kubernetes-controller" \
+                --loglevel="${LOGLEVEL}"
+        env:
+          - name: LOGLEVEL
+            value: "5"
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply bindata/network/node-identity/managed/node-identity.yaml]: set to 2 as in the original patch

File: bindata/network/node-identity/self-hosted/node-identity-webhook.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: network-node-identity.openshift.io
+webhooks:
+  - name: node.network-node-identity.openshift.io
```
[the remark this answers is not in the document]
Comment (JacobTanenbaum) [reply bindata/network/node-identity/self-hosted/node-identity-webhook.yaml:7]: added dd63d65a0be5b9ff97f86ff596727147f663fc16

File: bindata/network/ovn-kubernetes/common/002-rbac.yaml (hunk 3 above)
Comment (JacobTanenbaum) [reply bindata/network/ovn-kubernetes/common/002-rbac.yaml:24]: removed the permissions

**Pull request closed without merging**